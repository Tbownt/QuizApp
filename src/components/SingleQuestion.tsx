import { useEffect, useMemo, useState } from "react";

const questionLetters = ["A", "B", "C", "D"];

export const SingleQuestion = ({ question, onAnswerSubmit }) => {
  // Guarda la respuesta correcta

  const correctAnswer = useMemo(
    () => question.options.find((item) => item.valid === true),
    [question]
  );

  const [userSelectedAnswer, setUserSelectedAnswer] = useState(false);
  const [selectedAnswer, setSelectedAnswer] = useState(null);

  // Limpia la respuesta seleccionada cuando se cambia de pregunta
  useEffect(() => {
    setSelectedAnswer(null);
    setUserSelectedAnswer(false);
  }, [question]);

  const listAnswersByCapitalQuestion = (questionItem = []) => {
    if (!questionItem) return null;

    const correctAnswer = questionItem.options.find(
      (item) => item.valid === true
    );

    return (
      <>
        <p className="text-title font-bold w-96 text-2xl my-2">
          La capital de {correctAnswer.countryName} es:{" "}
        </p>
        <ul>
          {questionItem.options.map((item, index) => (
            <li
              style={{
                backgroundColor: selectedAnswer
                  ? getAnswerColors(item)
                  : "transparent",
              }}
              onClick={() => validateAnswer(item, correctAnswer)}
              key={`${index}-${item.capital}`}
              className="my-3 border-[2px] rounded-[6px] p-2 border-title"
            >
              <span className="text-title text-[18px]">
                {`${questionLetters[index]}.`} {item.capital}{" "}
              </span>
            </li>
          ))}
        </ul>
      </>
    );
  };

  const validateAnswer = (selectedOption, correctAnswer) => {
    if (userSelectedAnswer) return;

    setSelectedAnswer(selectedOption);
    setUserSelectedAnswer(true);

    if (selectedOption.countryName === correctAnswer.countryName) {
      onAnswerSubmit({
        correct: true,
        selectedAnswer: selectedOption,
        correctAnswer,
      });
      return;
    }
    onAnswerSubmit({
      correct: false,
      selectedAnswer: selectedOption,
      correctAnswer,
    });
  };

  const getAnswerColors = (questionItem) => {
    // Primero se valida si es una respuesta correcta, comparando el item iterado con la respuesta correcta.
    if (questionItem.valid === correctAnswer.valid) return "#60BF88";

    // Si el item actual no es la respuesta seleccionada, entonces se marca naranja.
    if (questionItem.countryName !== selectedAnswer.countryName)
      return "orange";

    // De lo contrario, si es una respuesta incorrecta, se marca rojo.
    if (selectedAnswer.valid !== correctAnswer.valid) return "#EA8282";
  };

  const listAnswersByFlagQuestion = (questionItem = []) => {
    if (!questionItem) return null;
    return (
      <>
        <div className="flex flex-col">
          <img
            src={correctAnswer.flag}
            alt={correctAnswer.flagAlt}
            className="w-[84px] h-[54px]"
          />
          <p className="text-title font-bold w-96 text-2xl my-2">
            A cual pa√≠s pertenece esta bandera?
          </p>
          <ul className="">
            {questionItem.options.map((item, index) => (
              <li
                style={{
                  backgroundColor: selectedAnswer
                    ? getAnswerColors(item)
                    : "transparent",
                }}
                className="my-3 border-[2px] rounded-[6px] p-2 border-title"
                onClick={() => validateAnswer(item, correctAnswer)}
                key={`${index}-${item.countryName}`}
              >
                <span className="text-title text-[18px]">
                  {`${questionLetters[index]}.`} {item.countryName}
                </span>
              </li>
            ))}
          </ul>
        </div>
      </>
    );
  };

  const displayAnswerListByQuestionType = (currentQuestionItem) => {
    if (currentQuestionItem.checkFor === "flag") {
      return <>{listAnswersByFlagQuestion(question)}</>;
    }

    if (currentQuestionItem.checkFor === "capital") {
      return <>{listAnswersByCapitalQuestion(question)}</>;
    }
  };

  return (
    <div className="question">{displayAnswerListByQuestionType(question)}</div>
  );
};
