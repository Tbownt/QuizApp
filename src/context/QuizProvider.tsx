import { useState } from "react";
import { QuizContext } from "./QuizContext";
import { getCountryGroupArray } from "../helpers/countriesApi";
import { generateQuestionArray } from "../helpers/questionHelper";

export const QuizProvider = ({ children }) => {
  const initialQuizStatus = {
    completed: false,
    currentQuestion: 1,
    correctAnswers: 0,
  };

  const [rawCountryData, setRawCountryData] = useState();
  const [questionArray, setQuestionArray] = useState();
  const [quizStatus, setQuizStatus] = useState(initialQuizStatus);

  const increaseCorrectAnswerCount = () => {
    setQuizStatus((state) => ({
      ...state,
      correctAnswers: state.correctAnswers + 1,
    }));
  };

  const goToNextQuestion = () => {
    setQuizStatus((state) => ({
      ...state,
      completed: state.currentQuestion === 10 ? true : false,
      currentQuestion:
        state.currentQuestion < 10
          ? state.currentQuestion + 1
          : state.currentQuestion,
    }));
  };

  const restartQuiz = () => {
    const countryGroupArray = getCountryGroupArray(rawCountryData);
    setQuestionArray(generateQuestionArray(countryGroupArray));

    setQuizStatus(initialQuizStatus);
  };

  return (
    <QuizContext.Provider
      value={{
        questionArray,
        quizStatus,
        setRawCountryData,
        setQuestionArray,
        increaseCorrectAnswerCount,
        goToNextQuestion,
        restartQuiz,
      }}
    >
      {children}
    </QuizContext.Provider>
  );
};
