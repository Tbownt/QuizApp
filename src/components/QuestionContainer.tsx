import { useContext, useState } from "react";
import { QuizContext } from "../context/QuizContext";
import { SingleQuestion } from "./SingleQuestion";

export const Question = ({ questions }) => {
  const { quizStatus, increaseCorrectAnswerCount, goToNextQuestion } =
    useContext(QuizContext);
  const [enableButtonNext, setEnableButtonNext] = useState(false);

  if (!questions) return null;

  const handlePageChange = () => {
    goToNextQuestion();
    setEnableButtonNext(false);
  };

  const handleUserClick = (response) => {
    if (response.correct) increaseCorrectAnswerCount();
    setEnableButtonNext(true);
  };

  return (
    <>
      <p className="text-title font-bold w-96 text-[16px]  text-center">
        Pregunta ({quizStatus.currentQuestion} / {questions.length})
      </p>
      {questions && (
        <SingleQuestion
          question={questions[quizStatus.currentQuestion - 1]}
          onAnswerSubmit={handleUserClick}
        />
      )}
      <div className="flex flex-row">
        <span className="text-title font-bold w-96 text-[16px]  text-left mt-3">
          Correctas ({quizStatus.correctAnswers} / {questions.length})
        </span>
        <div
          className={
            enableButtonNext
              ? "w-[35%] border text-center rounded-[12px] text-white bg-yellow h-10 self-end hover:border-title hover:border-[3px]"
              : ""
          }
        >
          {enableButtonNext && (
            <button onClick={handlePageChange} className="pt-[5px]">
              {quizStatus.currentQuestion === 10 ? "Resultados" : "Siguiente"}
            </button>
          )}
        </div>
      </div>
    </>
  );
};
