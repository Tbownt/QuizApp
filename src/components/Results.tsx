import { useContext } from "react";
import { QuizContext } from "../context/QuizContext";

export const Results = () => {
  const { quizStatus, restartQuiz } = useContext(QuizContext);
  return (
    <div className="flex flex-col items-center">
      <img
        src={"../assets/images/undraw_winners_ao2o2.svg"}
        alt="results_icon"
        className="w-[238px] flex-shrink-0 top"
      />
      <h2 className="text-5xl font-bold text-title mt-10">Resultado</h2>
      <p className="mt-5">
        Obtuviste{" "}
        <span className="text-2xl font-bold text-green">
          {quizStatus.correctAnswers}{" "}
        </span>
        respuestas correctas
      </p>
      <div className="border-2 mt-6 p-4 rounded-2xl">
        <button
          className="text-[18px] font-semibold text-title"
          onClick={restartQuiz}
        >
          Intentar de nuevo
        </button>
      </div>
    </div>
  );
};
