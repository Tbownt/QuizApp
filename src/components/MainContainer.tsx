import { useContext, useEffect } from "react";
import { useFetch } from "../hooks/useFetch";
import { QuizContext } from "../context/QuizContext";
import { getCountryGroupArray } from "../helpers/countriesApi";
import { generateQuestionArray } from "../helpers/questionHelper";
import { Question } from "./QuestionContainer";
import { Results } from "./Results";

export const MainContainer = () => {
  const { questionArray, setQuestionArray, quizStatus, setRawCountryData } =
    useContext(QuizContext);
  const { data, isLoading } = useFetch(
    "https://restcountries.com/v3.1/all?fields=name,flags,capital"
  );

  useEffect(() => {
    if (!isLoading && data) {
      setRawCountryData(data);
      const countryGroupArray = getCountryGroupArray(data);
      setQuestionArray(generateQuestionArray(countryGroupArray));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data]);

  return (
    <>
      {quizStatus.completed && <Results />}

      {!quizStatus.completed && <Question questions={questionArray} />}
    </>
  );
};
