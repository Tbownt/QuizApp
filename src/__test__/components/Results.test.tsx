import { render } from "@testing-library/react";
import { QuizContext } from "../../context/QuizContext";
import { Results } from "../../components/Results";
import { failedQuizContext, successfulQuizContext } from "../fixtures";

describe("Pruebas en <Results />", () => {
  beforeEach(() => jest.clearAllMocks());

  test("Debe de mostrar el resultado de un quiz completado con 10/10 preguntas correctas.", () => {
    const { container } = render(
      <QuizContext.Provider value={successfulQuizContext}>
        <Results />
      </QuizContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar el resultado de un quiz completado con 5/10 preguntas correctas.", () => {
    const { container } = render(
      <QuizContext.Provider value={failedQuizContext}>
        <Results />
      </QuizContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
