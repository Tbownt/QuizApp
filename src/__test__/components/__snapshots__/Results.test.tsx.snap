// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Results /> Debe de mostrar el resultado de un quiz completado con 5/10 preguntas correctas. 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <img
      alt="results_icon"
      class="w-[238px] flex-shrink-0 top"
      src="../assets/images/undraw_winners_ao2o2.svg"
    />
    <h2
      class="text-5xl font-bold text-title mt-10"
    >
      Resultado
    </h2>
    <p
      class="mt-5"
    >
      Obtuviste
       
      <span
        class="text-2xl font-bold text-green"
      >
        5
         
      </span>
      respuestas correctas
    </p>
    <div
      class="border-2 mt-6 p-4 rounded-2xl"
    >
      <button
        class="text-[18px] font-semibold text-title"
      >
        Intentar de nuevo
      </button>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <Results /> Debe de mostrar el resultado de un quiz completado con 10/10 preguntas correctas. 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <img
      alt="results_icon"
      class="w-[238px] flex-shrink-0 top"
      src="../assets/images/undraw_winners_ao2o2.svg"
    />
    <h2
      class="text-5xl font-bold text-title mt-10"
    >
      Resultado
    </h2>
    <p
      class="mt-5"
    >
      Obtuviste
       
      <span
        class="text-2xl font-bold text-green"
      >
        10
         
      </span>
      respuestas correctas
    </p>
    <div
      class="border-2 mt-6 p-4 rounded-2xl"
    >
      <button
        class="text-[18px] font-semibold text-title"
      >
        Intentar de nuevo
      </button>
    </div>
  </div>
</div>
`;
